pkg/cmd/app/translate.go

```diff
@@ -0,0 +1,151 @@
+)
+
+// maxLogOutputRaw is the maximum size we allow to allocate for logs before
+// being encoded to base64
+// See: https://stackoverflow.com/a/4715480/1100238
+var maxLogOutputRaw = int(math.Floor(float64(maxLogOutput)*3) / 4)
+
+//CfgData represents the data to be include in a configmap
+type CfgData struct {
+	Name       string
+	Status     string
+	Output     string
+	Repository string
+	Branch     string
+	Filename   string
+	Manifest   []byte
+	Icon       string
+}
+
+// TranslateConfigMap translates the app into a configMap
+func TranslateConfigMap(name string, data *CfgData) *apiv1.ConfigMap {
+	return &apiv1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: TranslateAppName(name),
+			Labels: map[string]string{
+				model.GitDeployLabel: "true",
+			},
+		},
+		Data: map[string]string{
+			nameField:     name,
+			statusField:   data.Status,
+			outputField:   base64.StdEncoding.EncodeToString([]byte(data.Output)),
+			repoField:     data.Repository,
+			branchField:   data.Branch,
+			filenameField: data.Filename,
+			yamlField:     base64.StdEncoding.EncodeToString(data.Manifest),
+			iconField:     data.Icon,
+		},
+	}
+}
+
+// SetStatus sets the status and output of a config map
+func SetStatus(cfg *apiv1.ConfigMap, status, output string) *apiv1.ConfigMap {
+	cfg.Data[statusField] = status
+	cfg.Data[outputField] = base64.StdEncoding.EncodeToString([]byte(output))
+	return cfg
+}
+
+// SetOutput sets the output of a config map
+func SetOutput(cfg *apiv1.ConfigMap, output string) *apiv1.ConfigMap {
+	cfg.Data[outputField] = base64.StdEncoding.EncodeToString([]byte(output))
+	return cfg
+}
+
+// UpdateOutput updates the configmap output with the logs
+func UpdateOutput(ctx context.Context, name, namespace string, output *bytes.Buffer, c kubernetes.Interface) error {
+	cmap, err := configmaps.Get(ctx, name, namespace, c)
+	if err != nil {
+		return err
+	}
+
+	// If the output is larger than the currentMaxLimit for the logs trim it.
+	// We can't really truncate the buffer since we would end up with an invalid json
+	// line for the last line, so we pick lines from the end while the line fits
+	var data []byte
+	if output.Len() > maxLogOutputRaw {
+		scanner := bufio.NewScanner(output)
+		linesInReverse := []string{}
+		for scanner.Scan() {
+			linesInReverse = append([]string{scanner.Text()}, linesInReverse...)
+		}
+		var cappedOutput string
+		var head string
+		for _, l := range linesInReverse {
+			head = l + "\n"
+			if len(cappedOutput)+len(head) > maxLogOutputRaw {
+				break
+			}
+			cappedOutput = head + cappedOutput
+		}
+		cappedOutput = strings.TrimPrefix(cappedOutput, head)
+		data = []byte(cappedOutput)
+	} else {
+		data = output.Bytes()
+	}
+
+	SetOutput(cmap, string(data))
+
+	return configmaps.Deploy(ctx, cmap, namespace, c)
+}
+
+//TranslateAppName translate the name into the pipeline name
+func TranslateAppName(name string) string {
+	return fmt.Sprintf("okteto-git-%s", name)
```

Comment:
This will be the same name that the backend is generating, right? Have we tested these changes with the backend? I have the feeling that we are overriding the content of the configmap when we translate the config map. If the config map is created by the backend and then updated by the deploy command, we are losing some important fields like `actionLock` or `actionName`. We would be losing also important labels like `dev.okteto.com/pipeline-name`, `dev.okteto.com/pipeline-namespace` or `dev.okteto.com/installer-running`. We have 2 options:

* When we run the deploy command, if the configmap already exists, we use that as a base for the Translate and we only override the fields we have to, but the rest are keeping the same. We have to use optimistic lock to update the config map to avoid issues between the backend and cli
* Use a different config map as we are doing with stacks. This would require changes in the backend to group together the configmaps when a pipeline is deployed.